src/main/java/com/commercetools/sync/services/impl/ProductServiceImpl.java

```diff
@@ -141,18 +142,9 @@ private void cacheProductIds(@Nonnull final List<Product> products) {
     @Nonnull
     @Override
     public CompletionStage<Set<Product>> createProducts(@Nonnull final Set<ProductDraft> productsDrafts) {
-        final List<CompletableFuture<Optional<Product>>> futureCreations =
-            productsDrafts.stream()
-                          .map(this::createProduct)
-                          .map(CompletionStage::toCompletableFuture)
-                          .collect(Collectors.toList());
-        return CompletableFuture.allOf(futureCreations.toArray(new CompletableFuture[futureCreations.size()]))
-                                .thenApply(result -> futureCreations.stream()
-                                                                    .map(CompletionStage::toCompletableFuture)
-                                                                    .map(CompletableFuture::join)
-                                                                    .filter(Optional::isPresent)
-                                                                    .map(Optional::get)
-                                                                    .collect(Collectors.toSet()));
+        return mapValuesSetToFutureOfCompletedValues(productsDrafts, this::createProduct)
+            .thenApply(results -> results.filter(Optional::isPresent).map(Optional::get))
```

Comment:
i found that approach `filter(Optional::isPresent).map(Optional::get)` is used very often...
Well, i voted already once to dismiss mostly Optionals in favor of Streams and flatMap, but even if we decided to stick with Optionals+Stream i'd recommend to carry out to some utils:

  - `filter(Optional::isPresent).map(Optional::get)`
  - `filter(Optional::isPresent).map(Optional::get).collect(listOrSetCollector)`

Reply:
I definitley agree, but also too much for this PR and time constraint of asset sync feature at the moment :( 

So issue created in the backlog: https://github.com/commercetools/commercetools-sync-java/issues/255